File: 백준/Gold/33714. Min Max Mex/Min Max Mex.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());

        ArrayList<Integer> nums = new ArrayList<>();
        HashMap<Integer, Integer> map = new HashMap<>();

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            int num = Integer.parseInt(st.nextToken());
            nums.add(num);

            if (map.containsKey(num)) {
                map.put(num, map.get(num) + 1);
            } else {
                map.put(num, 1);
            }
        }
        
        HashSet<Integer> set = new HashSet<>(nums);
        ArrayList<Integer> setToList = new ArrayList<>(set);
        setToList.sort(Comparator.naturalOrder());

        int min = N;
        for (int i = 0; i < N; i++) {
            if (!map.containsKey(i) || map.get(i) <= K) {
                min = i;
                break;
            }
        }

        System.out.println(min); // 최소값 출력


        int prev = -1; // 가능한 최솟값이 0이니 초기 prev는 -1
        boolean check = false;

        for (int i = 0; i < setToList.size(); i++) {
            // 이전 값과 현재 값 사이에 넣어야하는 숫자 갯수
            int fill = setToList.get(i) - prev - 1;

            if (fill == 0) { // 채울 필요가 없으면 패스
                prev = setToList.get(i);
            }

            // K번의 연산으로 채울 수 없다면 break
            if (K - fill < 0) {
                check = true;
                break;
            } else { // 채울 수 있다면 다음 값을 넘어간다
                K -= fill;
                prev = setToList.get(i);
            }
        }

        // break 되었다면 이전 값에 (K+1)
        if (check) {
            System.out.println(prev + (K+1));
        } else { // 아니라면 마지막 값에 K+1
            System.out.println(setToList.get(setToList.size() - 1)+ (K+1));
        }
    }
}

File: 백준/Gold/33714. Min Max Mex/Min Max Mex.py
# Min Max Mex
import sys
input = sys.stdin.readline

# Mex(A) -> 배열 A에 포함되어 있지 가장 작은 음이 아닌 정수
N, K = map(int,input().split())

nums = list(map(int,input().split()))
nums.sort()

sett = set(nums)
set_list = list(sett)
set_list.sort()

dic = {}

for i in range(N):
    num = nums[i]
    
    if num in dic:
        dic[num] += 1
    else:
        dic[num] = 1

# 가능한 작은 수는 최대 N
minn = N
for i in range(N):
    # dic에 없거나 제거할 수 있다면 해당 숫자가 최솟값
    if i not in dic or dic[i] <= K:
        minn = i
        break

print(minn)

prev = -1
check = False
for i in range(len(set_list)):
    fill = set_list[i] - prev - 1
    
    if fill < 1:
        prev = set_list[i]
        continue
    
    if K - fill < 0:
        check = True
        break
    else:
        K -= fill
        prev = set_list[i]

if check:
    print(prev + (K+1))
else:
    print(set_list[-1] + (K+1))

